OpenRA.Game/MiniYaml.cs

```diff
@@ -215,64 +245,81 @@ static List<MiniYamlNode> FromLines(IEnumerable<string> lines, string filename,
+					var prevChar = '\0';
+					for (var i = keyStart; i < lineLength; i++)
 					{
-						if (valueStart < 0 && line[i] == ':')
+						var currChar = line[i];
+						if (currChar == '#')
+						{
+							if (prevChar == '\\')
+								hasEscapedHashChar = true;
+							else
+							{
+								commentStart = i + 1;
+								if (commentStart <= keyLength)
+									keyLength = i - keyStart;
+								else
+									valueLength = i - valueStart;
+
+								break;
+							}
+						}
+						else if (valueStart < 0 && currChar == ':')
 						{
 							valueStart = i + 1;
 							keyLength = i - keyStart;
-							valueLength = line.Length - i - 1;
+							valueLength = lineLength - i - 1;
 						}
 
-						if (commentStart < 0 && line[i] == '#' && (i == 0 || line[i - 1] != '\\'))
-						{
-							commentStart = i + 1;
-							if (commentStart <= keyLength)
-								keyLength = i - keyStart;
-							else
-								valueLength = i - valueStart;
-
-							break;
-						}
+						prevChar = currChar;
 					}
 
 					if (keyLength > 0)
+					{
 						key = line.Substring(keyStart, keyLength).Trim();
+						key = stringPool.GetOrAdd(key, key);
+					}
 
 					if (valueStart >= 0)
 					{
 						var trimmed = line.Substring(valueStart, valueLength).Trim();
-						if (trimmed.Length > 0)
+						valueLength = trimmed.Length;
+						if (valueLength > 0)
+						{
 							value = trimmed;
+
+							if (valueLength > 1)
+							{
+								// Remove leading/trailing whitespace guards
+								var trimLeading = value[0] == '\\'
+									&& (value[1] == ' ' || value[1] == '\t') ? 1 : 0;
+								var trimTrailing = value[value.Length - 1] == '\\'
+									&& (value[valueLength - 2] == ' ' || value[valueLength - 2] == '\t') ? 1 : 0;
+								if (trimLeading + trimTrailing > 0)
+									value = value.Substring(trimLeading, valueLength - trimLeading - trimTrailing);
+							}
+
+							if (hasEscapedHashChar)
+								value = value.Replace("\\#", "#");
+
+							value = stringPool.GetOrAdd(value, value);
+						}
 					}
 
 					if (commentStart >= 0 && !discardCommentsAndWhitespace)
-						comment = line.Substring(commentStart);
-
-					// Remove leading/trailing whitespace guards
-					if (value != null && value.Length > 1)
 					{
-						var trimLeading = value[0] == '\\' && (value[1] == ' ' || value[1] == '\t') ? 1 : 0;
-						var trimTrailing = value[value.Length - 1] == '\\' && (value[value.Length - 2] == ' ' || value[value.Length - 2] == '\t') ? 1 : 0;
-						if (trimLeading + trimTrailing > 0)
-							value = value.Substring(trimLeading, value.Length - trimLeading - trimTrailing);
+						comment = line.Substring(commentStart);
+						comment = stringPool.GetOrAdd(comment, comment);
 					}
-
-					// Remove escape characters from #
-					if (value != null && value.IndexOf('#') != -1)
-						value = value.Replace("\\#", "#");
 				}
 
 				if (key != null || !discardCommentsAndWhitespace)
 				{
-					key = key == null ? null : stringPool.GetOrAdd(key, key);
```

Comment:
Please keep these here, where it is clear that they are always going to be called before adding the node to the tree. Moving these into the spaghetti above makes it far too easy for a future change to accidentally introduce an unpooled string into the tree.